plugin/core/windows.py

```diff
@@ -331,24 +327,20 @@ def _on_project_switched(self, folders: List[str]) -> None:
         self.end_sessions()
 
     def get_session(self, config_name: str, file_path: str) -> Optional[Session]:
-        try:
-            return self._find_session(config_name, file_path)
-        except InitializeError as ex:
-            assert ex.session.config.name == config_name
-            self._disable_temporarily(ex.session, ex, self._workspace.designated(file_path))
-        return None
+        return self._find_session(config_name, file_path)
+
+    def _is_session_ready(self, config_name: str, file_path: str) -> bool:
+        maybe_session = self._find_session(config_name, file_path)
+        return maybe_session is not None and maybe_session.state == ClientStates.READY
+
+    def _can_start_config(self, config_name: str, file_path: str) -> bool:
+        return not bool(self._find_session(config_name, file_path))
 
     def _find_session(self, config_name: str, file_path: str) -> Optional[Session]:
-        if file_path in self._workspace:
-            sessions = self._sessions.get(config_name, [])
-            for session in sessions:
+        if config_name in self._sessions:
+            for session in self._sessions[config_name]:
                 if session.handles_path(file_path):
                     return session
-        else:
-            sessions = self._sessions.get(config_name, [])
-            if not sessions:
-                return None
-            return sessions[0]
```

Comment:
I think common problems are that people have different expectations for code outside their project - for example we ignore diagnostics for these files but realistically why should the server even start compiling/syntax-checking or linting them?
The support that existed has been mostly bug-driven - people commenting "I think it should work like this for the server I'm using". I'd like to know what the spec / other editors do and add support for LSP handling these views as unusual lsp documents as soon as they are opened.

Reply:
Sounds good to me.

However, we need to think carefully about servers that don't care about workspaces at all (json/yaml/xml language servers) as those can run with a file anywhere on the file system.

We also need to think carefully about that annoying case where you can start sublime text with a single file and the `window.folders()` will be empty. Perhaps LSP can ignore this case as a user will probably not start doing serious work in this situation.

Reply:
To me running checking on files outside of the workspace is mostly annoying. Especially with linting servers like `eslint` for example. Opening a file outside of the workspace typically results in most of the file being annotated with errors as then eslint can't find proper linting configuration for that file and either applies some defaults or user's default which typically has different style expectations.

Reply:
I agree with @rwols. I think it would be nice to have a per-server  setting like `"useOnlyWithinWorkspace": false/true`, so the servers that are commonly used inside the workspace folder (js/ts/eslint etc) will not be annoying. And servers that don't care about workspaces at all (json/yaml/xml and other configs that can be opened from anywhere at any time) will continue to work as people expect.